### PR TITLE
Surface error for plain return statement in coroutine earlier

### PR DESCRIPTION
**Pull request by ivanaivanovska**
When a plain return statement was used in a coroutine, the error  "return statement not allowed in coroutine" was surfaced too late (e.g. after other errors in the return statement). Surfacing it earlier now, to make the issue more obvious.

### CODE REVIEW

File: clang/test/SemaCXX/coroutines.cpp
Comment (ilya-biryukov) [inline clang/test/SemaCXX/coroutines.cpp:299]: NIT: for brevity we could use a shorter substring (same below)
```suggestion
struct Handle : std::coroutine_handle<promise_handle> { // expected-note 2{{not viable}}
```

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaCoroutine.cpp:1123]: We now have some asymmetry between the way we handle errors for `co_await` after `return` and for `return` after `co_await`. (Also for other coroutine keywords, but I'll keep only mentioning `co_await` for simplicity).

I think we could get rid of that by moving this code into the function that handles `co_await`. It can check `FirstReturnLoc.isValid()` and issue and error in the same way as `return` checks for `FirstCoroutineStmtLoc.isValid()` (which is the implementation of `Fn->isCoroutine()`) and issues an error.

File: clang/lib/Sema/SemaStmt.cpp
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaStmt.cpp:3752]: Could we only do this when `FSI->FirstReturnLoc.isInvalid() == true` to make sure we do not start producing the error on every return statement?

Previously we only produced a single error, and this seems like a useful behavior to keep. Otherwise we're spamming people with error messages.

File: clang/test/SemaCXX/coroutines.cpp
Comment (ilya-biryukov) [inline clang/test/SemaCXX/coroutines.cpp:317]: the `-verify` and `expected-error` does not actually let us check the order in which the errors are reported.

Could we add another test file that would do something like:

```
// RUN: %clang_cc1 -fsyntax-only %s 2>& 1 | FileCheck
// The error about a function not being a coroutine is usually more relevant than the conversion error.
// Check that we show it first.
<test code here>
// CHECK-NOT: cannot convert
// CHECK: return statement not allowed in coroutine
// CHECK: cannot convert
```
it does not have to be detailed, just to make sure we actually managed to keep the errors in that order

**Review by ilya-biryukov — COMMENTED** [review]
Thanks for the change!
This looks like an improvement, but I suspect we can do even better. See the comments I've left in the code.

File: clang/test/SemaCXX/coroutines.cpp
Comment (ivanaivanovska) [reply clang/test/SemaCXX/coroutines.cpp]: Done.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ivanaivanovska) [reply clang/lib/Sema/SemaCoroutine.cpp]: Done.

File: clang/lib/Sema/SemaStmt.cpp
Comment (ivanaivanovska) [reply clang/lib/Sema/SemaStmt.cpp]: Done.

File: clang/test/SemaCXX/coroutines.cpp
Comment (ivanaivanovska) [reply clang/test/SemaCXX/coroutines.cpp:341]: Added tests for the order of the errors in the original file (coroutines.cpp) to keep the tests together and reuse the code.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaCoroutine.cpp:690]: NIT: negate the condition to reduce nesting, see LLVM style guide on [using early exits](https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code)

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaCoroutine.cpp:709]: NIT: remove braces from simple statements, see [LLVM Style Guide](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

File: clang/test/SemaCXX/coroutines.cpp
Comment (ilya-biryukov) [inline clang/test/SemaCXX/coroutines.cpp:342]: Could you add a comment explaining that we are testing the order of diagnostics here.

```
// Check that we first show that return is not allowed in coroutine, the error about bad conversion is
// most likely spurious so we prefer to have it afterwards.
```

File: clang/test/SemaCXX/coroutines.cpp
Comment (ilya-biryukov) [inline clang/test/SemaCXX/coroutines.cpp:6]: Could you add a comment here explaining to the reader why we need this? It's unusual to have `-verify` on some `RUN` lines and not have it on others.

Something like

```
// Run without -verify to check the order of errors we show.
// RUN: not ...
```

File: clang/test/SemaCXX/coroutines.cpp
Comment (ilya-biryukov) [reply clang/test/SemaCXX/coroutines.cpp:341]: LG, thanks!

File: clang/test/SemaCXX/coroutines.cpp
Comment (ilya-biryukov) [inline clang/test/SemaCXX/coroutines.cpp:350]: Could you run `clang-format` over these lines? (E.g. via `git-clang-format` or "Format Lines" action from VSCode with Clangd installed). The `// expected-error@-1` should get nicely aligned with the previous lines.

File: clang/test/SemaCXX/coroutines.cpp
Comment (ilya-biryukov) [inline clang/test/SemaCXX/coroutines.cpp:362]: NIT: could you add a comment explaining that we do not want the duplicate error?
(it's easy for the reader to understand the expected warnings that are written in the code, but it's harder to understand the intention when it's not written)

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaCoroutine.cpp:691]: NIT: we could move this assertion to the start of the function. As the name implies, the `checkReturnStmtInCoroutine` can only be called on coroutines, so validating this invariant is useful even if we don't report any diagnostics.

**Review by ilya-biryukov — COMMENTED** [review]
I have a lot of nitpicks, but otherwise LG, happy to approve as soon as they're fixed.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ivanaivanovska) [reply clang/lib/Sema/SemaCoroutine.cpp]: Done.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ivanaivanovska) [reply clang/lib/Sema/SemaCoroutine.cpp]: Done.

File: clang/test/SemaCXX/coroutines.cpp
Comment (ivanaivanovska) [reply clang/test/SemaCXX/coroutines.cpp:8]: Done.

File: clang/test/SemaCXX/coroutines.cpp
Comment (ivanaivanovska) [reply clang/test/SemaCXX/coroutines.cpp:346]: Done.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ivanaivanovska) [reply clang/lib/Sema/SemaCoroutine.cpp]: Done.

File: clang/test/SemaCXX/coroutines.cpp
Comment (ivanaivanovska) [reply clang/test/SemaCXX/coroutines.cpp]: Done.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaCoroutine.cpp:690]: This should be `assert(FSI)`, it should fail with the current version.

BTW, are you building with `LLVM_ENABLE_ASSERTIONS=On`? It's definitely recommended for the day-to-day development.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaCoroutine.cpp:710]: NIT: add a comment explaning why we don't do this every time:

```cpp
// Avoid duplicate errors, report only on first keyword.
```

**Review by ilya-biryukov — CHANGES_REQUESTED** [review]
The `assert` condition seems wrong, otherwise this should be good to go.

I've also left one more suggestion to leave a comment that I missed in the original review since we need another round anyway.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ivanaivanovska) [reply clang/lib/Sema/SemaCoroutine.cpp]: Right. No, assertions were off, so I missed it. I fixed that now. Thanks.

File: clang/lib/Sema/SemaCoroutine.cpp
Comment (ivanaivanovska) [reply clang/lib/Sema/SemaCoroutine.cpp:711]: Done.

**Review by ilya-biryukov — APPROVED** [review]
Thanks! The changes LG, I'll wait until premerge checks finish and merge it into mainline.

**Pull request merged**
